# OTP Verification System

## 1. Generate 6-digit OTP

In [1]:
import random

def generate_otp():
    """Generates a random 6-digit OTP."""
    return str(random.randint(100000, 999999))

# Demo the function
otp = generate_otp()
print(f"Generated OTP: {otp}")

Generated OTP: 967940


## 2. Simulate Sending OTP to Email

In [2]:
def send_otp_email(email_address, otp):
    """Simulates sending the OTP to the specified email address."""
    print(f"Simulating: Sending OTP {otp} to {email_address}")
    # In a real application, this would integrate with an email sending service

# Demo the function
user_email = "user@example.com"
send_otp_email(user_email, otp)

Simulating: Sending OTP 967940 to user@example.com


## 3. Prompt User for OTP Input

In [3]:
def get_user_otp_input():
    """Prompts the user to enter the received OTP."""
    return input("Please enter the 6-digit OTP you received: ")

# Demo the function (we'll integrate this with verification later)
# user_entered_otp = get_user_otp_input()
# print(f"User entered: {user_entered_otp}")

## 4. Verify OTP and Handle Retries

In [4]:
def verify_otp(generated_otp, max_retries=3):
    """Verifies if the entered OTP matches the generated OTP, allowing retries."""
    for attempt in range(max_retries):
        user_entered_otp = get_user_otp_input()
        if user_entered_otp == generated_otp:
            print("OTP verified successfully! Access granted.")
            return True
        else:
            print(f"Incorrect OTP. {max_retries - 1 - attempt} attempts remaining.")
    print("Too many incorrect attempts. Access denied.")
    return False

# Demo the full process (assuming otp is already generated from previous cell)
print("\n--- OTP Verification Process ---")
send_otp_email(user_email, otp)
verification_status = verify_otp(otp)
print(f"Verification Status: {verification_status}")


--- OTP Verification Process ---
Simulating: Sending OTP 967940 to user@example.com
Please enter the 6-digit OTP you received: 967940
OTP verified successfully! Access granted.
Verification Status: True


## 5. Documentation

### Project Title: OTP Verification System

### Functionality Overview:
This system generates a 6-digit One-Time Password (OTP), simulates sending it to a user's email, and then verifies the user's input against the generated OTP with a limited number of retries.

### Dependencies:
This script uses standard Python libraries. No external dependencies are required beyond a basic Python installation.

### Functions Implemented:

1.  `generate_otp()`
    *   **Description:** Generates a random 6-digit numeric string to be used as an OTP.
    *   **Returns:** A string representing the 6-digit OTP.

2.  `send_otp_email(email_address, otp)`
    *   **Description:** Simulates the process of sending an OTP to a specified email address. In a real-world scenario, this function would integrate with an email service provider (e.g., SendGrid, Mailgun) to dispatch the email.
    *   **Parameters:**
        *   `email_address` (str): The recipient's email address.
        *   `otp` (str): The OTP to be sent.
    *   **Returns:** None (prints a simulated message).

3.  `get_user_otp_input()`
    *   **Description:** Prompts the user to enter the OTP they received via their simulated email. It captures and returns the user's input.
    *   **Returns:** A string representing the OTP entered by the user.

4.  `verify_otp(generated_otp, max_retries=3)`
    *   **Description:** Compares the user-entered OTP with the system-generated OTP. It allows for a configurable number of retries (defaulting to 3) before denying access.
    *   **Parameters:**
        *   `generated_otp` (str): The OTP that was originally generated by the system.
        *   `max_retries` (int, optional): The maximum number of attempts the user has to enter the correct OTP. Defaults to 3.
    *   **Returns:** `True` if the OTP is successfully verified, `False` otherwise.

### How to Run the Program:

1.  **Ensure Python is installed:** This script is written in Python and requires a Python interpreter (version 3.6 or newer recommended).
2.  **Run the cells in sequence:** Execute each code cell in the Colab notebook from top to bottom. The demonstration in the final `verify_otp` cell will guide you through the process.
    *   The `generate_otp` cell will generate and display an OTP.
    *   The `send_otp_email` cell will simulate sending this OTP to a placeholder email address.
    *   The final `verify_otp` cell will then prompt you to enter the OTP. You should enter the OTP that was displayed in the first cell's output. If you enter an incorrect OTP, you will be prompted to try again up to `max_retries` times.

### Example Workflow:

1.  **Execute `generate_otp()`:** This will print `Generated OTP: XXXXXX` (e.g., `967940`). Remember this OTP.
2.  **Execute `send_otp_email()`:** This will print `Simulating: Sending OTP XXXXXX to user@example.com`.
3.  **Execute `verify_otp()`:**
    *   The system will prompt: `Please enter the 6-digit OTP you received:`
    *   **To succeed:** Enter the exact OTP (e.g., `967940`). Output: `OTP verified successfully! Access granted.`
    *   **To fail:** Enter an incorrect OTP. You will see `Incorrect OTP. Y attempts remaining.` After `max_retries` incorrect attempts, it will print `Too many incorrect attempts. Access denied.`
